core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -655,6 +655,54 @@ public int size() {
         return ret;
       };
 
+
+  /**
+   * Type-inference strategy for String concatenation.
+   * For example,
+   *
+   * concat(cast('a' as varchar(2)), cast('b' as varchar(3)),cast('c' as varchar(2)))
+   * returns varchar(7),
+   *
+   * concat(cast('a' as varchar), cast('b' as varchar(2), cast('c' as varchar(2))))
+   * returns varchar,
+   *
+   * concat(cast('a' as varchar(65535)), cast('b' as varchar(2)), cast('c' as varchar(2)))
+   * returns varchar
+   */
+  public static final SqlReturnTypeInference MULTIVALENT_STRING_SUM_PRECISION =
+      opBinding -> {
+        boolean hasPrecisionNotSpecifiedOperand = false;
+        int typePrecision;
+        long amount = 0;
+        List<RelDataType> operandTypes = opBinding.collectOperandTypes();
+        for (RelDataType operandType: operandTypes) {
```

Comment:
You can directly break the iteration when: 1) encountering `PRECISION_NOT_SPECIFIED`, 2) `amount` large than max precision.